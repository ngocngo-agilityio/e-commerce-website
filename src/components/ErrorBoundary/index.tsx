'use client';

// Libs
import { useEffect } from 'react';
import { Button, Heading, Text, VStack } from '@chakra-ui/react';

// Constants
import { FOOTER_HEIGHT, HEADER_HEIGHT } from '@constants';

interface ErrorProps {
  error: Error & { digest?: string };
  reset: () => void;
}

const ErrorBoundary = ({ error, reset }: ErrorProps) => {
  // Attempt to recover by trying to re-render the segment
  const handleReset = () => {
    reset();
  };

  useEffect(() => {
    // Log the error to an error reporting service
    console.error(error);
  }, [error]);

  return (
    <VStack
      justifyContent="center"
      gap="50px"
      minHeight={`calc(100vh - ${HEADER_HEIGHT} - ${FOOTER_HEIGHT})`}
    >
      <Heading>Something went wrong!</Heading>
      <Text color="red">{error?.message}</Text>
      <Button onClick={handleReset}>Try again</Button>
    </VStack>
  );
};

export default ErrorBoundary;
