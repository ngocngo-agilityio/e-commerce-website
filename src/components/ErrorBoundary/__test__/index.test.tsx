import { fireEvent, render } from '@testing-library/react';

// Components
import ErrorBoundary from '..';

const mockProps = {
  error: new Error('Test Error'),
  reset: jest.fn(),
};

describe('ErrorBoundary Component', () => {
  it('Should render ErrorBoundary Component correctly', () => {
    const component = render(<ErrorBoundary {...mockProps} />);

    expect(component).toMatchSnapshot();
  });

  it('renders error message and calls reset function when  button is clicked', () => {
    const { getByText } = render(<ErrorBoundary {...mockProps} />);

    const tryAgainButton = getByText('Try again');
    fireEvent.click(tryAgainButton);

    expect(mockProps.reset).toHaveBeenCalled();
  });
});
