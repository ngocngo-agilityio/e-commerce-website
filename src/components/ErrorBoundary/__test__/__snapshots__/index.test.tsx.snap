// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary Component Should render ErrorBoundary Component correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="chakra-stack css-1ts0oy9"
      >
        <h2
          class="chakra-heading css-0"
        >
          Something went wrong!
        </h2>
        <p
          class="chakra-text css-tokvmb"
        >
          Test Error
        </p>
        <button
          class="chakra-button css-4xx2wk"
          type="button"
        >
          Try again
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="chakra-stack css-1ts0oy9"
    >
      <h2
        class="chakra-heading css-0"
      >
        Something went wrong!
      </h2>
      <p
        class="chakra-text css-tokvmb"
      >
        Test Error
      </p>
      <button
        class="chakra-button css-4xx2wk"
        type="button"
      >
        Try again
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
